core/go/internal/privatetxnmgr/transaction_flow_actions.go

```diff
@@ -370,7 +370,8 @@ func (tf *transactionFlow) writeAndLockStates(ctx context.Context) {
 	// we are accessing the transactionFlow's PrivateTransaction object which is only safe to do on the sequencer thread
 	// but we need to make sure that these writes/locks are complete before the assemble requester thread can proceed to assemble
 	// the next transaction
-	if tf.transaction.PostAssembly.OutputStatesPotential != nil && tf.transaction.PostAssembly.OutputStates == nil {
+	if (tf.transaction.PostAssembly.OutputStatesPotential != nil && tf.transaction.PostAssembly.OutputStates == nil) ||
```

Comment:
Spending a bit of time working out where this function fits into the transaction flow...

Reply:
Ok. This is adding the existence of states into the domain context. But these `info` states should never become available. So I'm not yet sure we're right here...

Reply:
No - it's all good. That critical difference between the two types of state (output states that become available, vs. info states that just need storing/distributing) is covered in the called function:
https://github.com/LF-Decentralized-Trust-labs/paladin/blob/5290e466f55c29848d2b28ec5165693c5ed77cfe/core/go/internal/domainmgr/private_smart_contract.go#L288

Reply:
Yes, this block is storing the state data for new states proposed in this assembled transaction. That should include both output states and info states.

Overall I'm a bit uncomfortable with the transaction flow here, where we iteratively look for `nil` things that we think we can populate. This fix has resolved a later panic as detailed in #502 because of some assumptions about what will and won't be `nil` at various points in the flow - and it's very difficult to follow how this check leads to that one.

It feels like there may be many more edge cases like this lurking around. I've not taken on a larger refactor at the moment, but it feels like one may be needed at some point.